DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs

```diff
@@ -232,10 +238,79 @@ public IActionResult ConfirmDeleteGroup(int groupId, ConfirmDeleteGroupViewModel
             return RedirectToAction("Index");
         }
 
+        [HttpGet("Generate")]
+        public IActionResult GenerateGroups()
+        {
+            var registrationFieldOptions = GetRegistrationFieldOptionsSelectList();
+
+            var model = new GenerateGroupsViewModel(registrationFieldOptions);
+            return View(model);
+        }
+
+        [HttpPost("Generate")]
+        public IActionResult GenerateGroups(GenerateGroupsViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                model.RegistrationFieldOptions = GetRegistrationFieldOptionsSelectList(model.RegistrationFieldOptionId);
+                return View(model);
+            }
+
+            var adminId = (int)User.GetAdminId()!;
+            var centreId = User.GetCentreId();
+
+            var groupDetails = new GroupGenerationDetails(
+                adminId,
+                centreId,
+                (int)model.RegistrationFieldOptionId!,
+                model.PrefixGroupName,
+                model.PopulateExisting,
+                model.AddNewRegistrants,
+                model.SyncFieldChanges,
+                model.SkipDuplicateNames
+            );
+
+            groupsService.GenerateGroupsFromRegistrationField(groupDetails);
+
+            return RedirectToAction("Index");
+        }
+
         private IEnumerable<CustomPrompt> GetRegistrationPromptsWithSetOptions(int centreId)
         {
             return centreCustomPromptsService.GetCustomPromptsForCentreByCentreId(centreId).CustomPrompts
                 .Where(cp => cp.Options.Any());
         }
+
+        private IEnumerable<SelectListItem> GetRegistrationFieldOptionsSelectList(int? selectedId = null)
+        {
+            var registrationFieldOptions = GetCustomPromptOptions();
+
+            var jobGroupOption = (7, "Job group");
+            registrationFieldOptions.Add(jobGroupOption);
+
+            return SelectListHelper.MapOptionsToSelectListItems(registrationFieldOptions, selectedId);
+        }
+
+        private List<(int id, string name)> GetCustomPromptOptions()

```

Comment:
If we're doing this mapping in the controller, we probably need to add some unit tests for it

Reply:
Good idea, I've moved it to there, and added controller tests to ensure the registration field options get populated correctly